Contract Updates

mds1 opened the pull request:
Main changes:
- Add transfer restrictions

Other changes:
- Removed unnecessary `Minted` and `Burned` events
- Update [from Buidler to Hardhat](https://medium.com/nomic-labs-blog/buidler-has-evolved-introducing-hardhat-4bccd13bc931)
- Update code comments
- Update linter/prettier settings

You can see the factory contract deployed and verified on Rinkeby here: (https://rinkeby.etherscan.io/address/0x4FB8Ad81754a5330165277a2655d47e99A7F598E)

mds1 committed:
Update lint/prettier config, run prettier

mds1 committed:
Update from buidler to hardhat

mds1 committed:
Add contract comments, remove unnecessary events

mds1 committed:
Implement transfer restrictions

apbendi commented on contracts/PToken.sol:
```diff
@@ -9,71 +9,139 @@ contract PToken is ERC20UpgradeSafe, OwnableUpgradeSafe {
+    // This contract can send tokens to anyone. If the sender is anyone except this contract, the
+    // only allowed recipient is this contract
+    if (_from != address(this)) {
+      require(_to == address(this), "PToken: Invalid recipient");
```
One minor nitpick— why use an if statement here? Why not just:

```solidity
require(_from != address(this) || _to == address(this), "PToken: Invalid recipient");
```

Or, more readably:

```solidity
bool isToOrFromContract = _from != address(this) || _to == address(this);
require(isToOrFromContract, "PToken: Invalid recipient");
```

mds1 replied:
Updated in 3dabd14

apbendi reviewed (approved):
Looks good! One minor stylistic nitpick— nice work though :)

mds1 committed:
Simplify transfer restriction  modifier

mds1 commented:
Before merging this, I'm going to deploy the final mainnet and Rinkeby versions. But I'm going to wait a few days to do that final deploy, because:
- We verify the contract on Etherscan using Hardhat's [Etherscan plugin](https://hardhat.org/plugins/nomiclabs-hardhat-etherscan.html)
- As you can see on our last [Rinkeby Etherscan deployment](https://rinkeby.etherscan.io/address/0x4FB8Ad81754a5330165277a2655d47e99A7F598E#code), file 12 of 12 is `MockDAI.sol`—it included an unrelated test file as part of the verification, which we don't want
- This is due to a [bug in solc](https://github.com/ethereum/solidity/issues/9573)
- Hardhat will have their own fix for this in [a few days](https://github.com/nomiclabs/hardhat/issues/804#issuecomment-728144991)

Because Etherscan verification is a pain, and we still need to finish a few other tasks before launch, I think it's ok to wait a few days until that's fixed before deploying the final contracts

mds1 committed:
WIP: Deploy contracts

mds1 committed:
Small contract update and final contract deployment

Removed the immutable keyword from the Factory contract, because
Etherscan would not verify the contract when present. Since the only
benefit of that keyword is slightly reduced gas costs when creating a
token, it was easiest to just remove that keyword

mds1 commented:
Mainnet and rinkeby contracts deployed and verified on Etherscan. Both are at [0x358bcf43fe7ec2659aD829F3604c72781fc93a9E](https://etherscan.io/address/0x358bcf43fe7ec2659aD829F3604c72781fc93a9E)
